### PR TITLE
Shard data extraction

### PR DESCRIPTION
**Pull request by SteveJRose**
Includes current shard counts, last hero pulled (not sure how useful that will be) and the pity system.

### CODE REVIEW

File: RaidExtractor.Core/AccountDumpClient.cs [hunk 1]
```diff
@@ -452,7 +455,30 @@ public partial class Hero
...
+    public partial class ShardInfo
+    {
+        [Newtonsoft.Json.JsonProperty("count", Required = Newtonsoft.Json.Required.Always)]
+        public int Count { get; set; }
+
+        [Newtonsoft.Json.JsonProperty("summonData", Required = Newtonsoft.Json.Required.Always)]
+        public List<ShardSummonInfo> SummonData { get; set; }
+    }
+
+        [System.CodeDom.Compiler.GeneratedCode("NSwag", "13.7.0.0 (NJsonSchema v10.1.24.0 (Newtonsoft.Json v11.0.0.0))")]
```
Comment (LukeCroteau) [inline RaidExtractor.Core/AccountDumpClient.cs:481]: Indentation is a little off on line 481 here

File: RaidExtractor.Core/Extractor.cs [hunk 2]
```diff
@@ -371,12 +371,100 @@ public AccountDump GetDump()
...
+                NativeWrapper.ReadProcessMemory(handle, shardsDataPointer + RaidStaticInformation.ListCount, ref shardCount);
+
+                List<IntPtr> shardPointers = new List<IntPtr>();
+                if (shardCount > 0)
+                {
+                    var arrayPointer = shardsDataPointer;
+                    NativeWrapper.ReadProcessMemory(handle, shardsDataPointer + 0x10, ref arrayPointer); 
+
+                    var ptrs = new IntPtr[shardCount];
+                    NativeWrapper.ReadProcessMemoryArray(handle, arrayPointer + 0x20, ptrs);
```
Comment (LukeCroteau) [inline RaidExtractor.Core/Extractor.cs:397]: Are these 0x20s just the standard List.Count, or is that 0x18?
Can we de-magic-number these? (Adding this comment for all the ones I see)
If not, it's ok. It's just a goal. I can re-factor it later.

File: RaidExtractor.Core/Extractor.cs [hunk 3]
```diff
@@ -371,12 +371,100 @@ public AccountDump GetDump()
...
+                NativeWrapper.ReadProcessMemory(handle, shardSummonDataPointer + RaidStaticInformation.ListCount, ref shardSummonCount);
+
+                shardPointers = new List<IntPtr>();
+                if (shardSummonCount > 0)
+                {
+                    var arrayPointer = shardSummonDataPointer;
+                    NativeWrapper.ReadProcessMemory(handle, shardSummonDataPointer + 0x10, ref arrayPointer);
+
+                    var ptrs = new IntPtr[shardSummonCount];
+                    NativeWrapper.ReadProcessMemoryArray(handle, arrayPointer + 0x20, ptrs);
```
Comment (LukeCroteau) [inline RaidExtractor.Core/Extractor.cs:430]: Are these 0x20s just the standard List.Count, or is that 0x18?
Can we de-magic-number these? (Adding this comment for all the ones I see)
If not, it's ok. It's just a goal. I can re-factor it later.

File: RaidExtractor.Core/Extractor.cs [hunk 4]
```diff
@@ -371,12 +371,100 @@ public AccountDump GetDump()
...
+
+                // Now read the summon data (pity system)
+                var shardSummonCount = 0;
+                NativeWrapper.ReadProcessMemory(handle, shardSummonDataPointer + RaidStaticInformation.ListCount, ref shardSummonCount);
+
+                shardPointers = new List<IntPtr>();
+                if (shardSummonCount > 0)
+                {
+                    var arrayPointer = shardSummonDataPointer;
+                    NativeWrapper.ReadProcessMemory(handle, shardSummonDataPointer + 0x10, ref arrayPointer);
```
Comment (LukeCroteau) [inline RaidExtractor.Core/Extractor.cs:427]: Can we de-magic-number these? (Adding this comment for all the ones I see)
If not, it's ok. It's just a goal. I can re-factor it later.

File: RaidExtractor.Core/Extractor.cs [hunk 5]
```diff
@@ -371,12 +371,100 @@ public AccountDump GetDump()
...
+                NativeWrapper.ReadProcessMemory(handle, shardsDataPointer + RaidStaticInformation.ShardData, ref shardsDataPointer); 
+
+                var shardCount = 0;
+                NativeWrapper.ReadProcessMemory(handle, shardsDataPointer + RaidStaticInformation.ListCount, ref shardCount);
+
+                List<IntPtr> shardPointers = new List<IntPtr>();
+                if (shardCount > 0)
+                {
+                    var arrayPointer = shardsDataPointer;
+                    NativeWrapper.ReadProcessMemory(handle, shardsDataPointer + 0x10, ref arrayPointer);
```
Comment (LukeCroteau) [inline RaidExtractor.Core/Extractor.cs:394]: Can we de-magic-number these? (Adding this comment for all the ones I see)
If not, it's ok. It's just a goal. I can re-factor it later.

**Review by LukeCroteau — CHANGES_REQUESTED** [review]
Nothing critical here.
Indentation is a minor code smell to me, if you could check and make sure that there's not a closed bracket in there and fix the indent on that line, I'd appreciate it.

As for the magic numbers, we *might* have a lot of 0x10 and 0x20 hanging around, and probably just need to tie them together. If you can't, it's totally fine, but let me know! Thanks!

**Comment by LukeCroteau** [comment]
And just as a regular comment, I appreciate this PR! This is cool functionality and I love it!

File: RaidExtractor.Core/AccountDumpClient.cs (hunk 1 above)
Comment (SteveJRose) [reply RaidExtractor.Core/AccountDumpClient.cs]: Arg! missed one, thanks.

File: RaidExtractor.Core/Extractor.cs (hunk 2 above)
Comment (SteveJRose) [reply RaidExtractor.Core/Extractor.cs]: de-magicified them: also changed the same numbers in pre-existing artifact extraction code .

**Comment by SteveJRose** [comment]
I've done some more testing today, and I'm seeing some strange results. I'm polling this until I know what's going on.

**Comment by SteveJRose** [comment]
Reopening. Shard data is held server side, and is only updated on app startup, which confused me. The code is correct.

**Review by LukeCroteau — APPROVED** [review]
Looks wonderful! Thanks so much for going back through and wiping Magic Numbers!

**Pull request merged**